### ISSUE TITLE
CFE inference bug: redirecting const factory constructors

### ISSUE THREAD

**Issue by sigmundch** [dart-lang/sdk#33813]
When using redirecting factory const constructors, inferred type-parameters seem incorrect.

Consider these 2 programs **v1.dart**:
```dart
class _X<T> {
  const factory _X() = _Y<T>;
}

class _Y<T> implements _X<T>{
  const _Y();
}

class A<T> {
  _X<T> x;
  A(this.x);
}

class B<T> extends A<T> {
  B(): super(const _X());
}

main() {
 print(new B().x.runtimeType);
}
```
and **v2.dart**:
```dart
class _Y<T> {
  const _Y();
}

class A<T> {
  _X<T> x;
  A(this.x);
}

class B<T> extends A<T> {
  B(): super(const _Y());
}

main() {
 print(new B().x.runtimeType);
}
```

In v1.dart, we print `_Y<Null>`, in v2.dart this produces an error because we try to create `_Y<B.T>`.

You can reproduce this on the VM (where you get an error in v2) and dart2js (where you get a compiler crash because we didn't expect such input).

Note this was working correctly in -dev.67. I'm seeing this issue in internal customer apps using the latest code in master. Their code contains uses of `Stream.empty`, similar to the code below:
```dart
import 'dart:async';

class A<T> {
  Stream<T> x;
  A(this.x);
}

class B<T> extends A<T> {
  B(): super(const Stream.empty());
}
```

@mit-mit could you help us triage?

**Comment by stefantsov** [comment]
Thanks for filing it, @sigmundch.  It's a long-standing one.  I've recently landed a pre-requisite for the fix: https://github.com/dart-lang/sdk/issues/32130.  Going to look at this one tomorrow and fix soon.

**Comment by sigmundch** [comment]
Thanks Dima, in case it is relevant, I did a bisect and noticed that that same commit aa6dc1e2d5081a23ee9d23e80a57237f9426f158 to fix #32130 is where this problem started to show up.

**Comment by stefantsov** [comment]
So, I initially thought that it's an instance of https://github.com/dart-lang/sdk/issues/32049 which is a known issue with type inference and redirecting factory constructors.  But I've come to the conclusion that it's not quite what I though it is originally.

My understanding of type inference is that the argument to `super` in both cases should be inferred as `B.T`, and the error should be reported in both examples.  Although, I'm not sure if it should be a compile-time error or a run-time error.  @leafpetersen Could you clarify how type inference should work here?

**Comment by sigmundch** [comment]
just an update on our end, we have a workaround for the issue, so I don't consider this a blocker anymore. I've removed the Dart2Stable milestone.

**Comment by leafpetersen** [comment]
> My understanding of type inference is that the argument to super in both cases should be inferred as B.T, and the error should be reported in both examples.

Not quite.  Inference is done exactly as with anything else, but then the inferred type arguments are replaced with their least closure.  So `B.T` is inferred, and then replaced with `Null` in each of them.  This is a valid static and runtime value to provide, so there are no errors.

**Comment by stefantsov** [comment]
Thanks for the clarification, Leaf!

I've created [CL 64841](https://dart-review.googlesource.com/c/sdk/+/64841) to fix the issue.  It appeared to be a one-liner that passes the `isConst` flag to the inference procedure.
